SDWebImage/SDWebImageDecoder.m

```diff
@@ -8,265 +8,124 @@
-        CGColorSpaceRef colorspaceRef = [UIImage colorSpaceForImageRef:imageRef];
-        
-        size_t width = CGImageGetWidth(imageRef);
-        size_t height = CGImageGetHeight(imageRef);
-        size_t bytesPerRow = kBytesPerPixel * width;
+@interface SDWebImageDecoder ()
 
-        // kCGImageAlphaNone is not supported in CGBitmapContextCreate.
-        // Since the original image here has no alpha info, use kCGImageAlphaNoneSkipLast
-        // to create bitmap graphics contexts without alpha info.
-        CGContextRef context = CGBitmapContextCreate(NULL,
-                                                     width,
-                                                     height,
-                                                     kBitsPerComponent,
-                                                     bytesPerRow,
-                                                     colorspaceRef,
-                                                     kCGBitmapByteOrderDefault|kCGImageAlphaNoneSkipLast);
-        if (context == NULL) {
-            return image;
-        }
-        
-        // Draw the image into the context and retrieve the new bitmap image without alpha
-        CGContextDrawImage(context, CGRectMake(0, 0, width, height), imageRef);
-        CGImageRef imageRefWithoutAlpha = CGBitmapContextCreateImage(context);
-        UIImage *imageWithoutAlpha = [UIImage imageWithCGImage:imageRefWithoutAlpha
-                                                         scale:image.scale
-                                                   orientation:image.imageOrientation];
-        
-        CGContextRelease(context);
-        CGImageRelease(imageRefWithoutAlpha);
-        
-        return imageWithoutAlpha;
-    }
-}
-
-/*
- * Defines the maximum size in MB of the decoded image when the flag `SDWebImageScaleDownLargeImages` is set
- * Suggested value for iPad1 and iPhone 3GS: 60.
- * Suggested value for iPad2 and iPhone 4: 120.
- * Suggested value for iPhone 3G and iPod 2 and earlier devices: 30.
- */
-static const CGFloat kDestImageSizeMB = 60.0f;
+@property (nonatomic, strong) id<SDWebImageCoder> incrementalCoder;
 
-/*
- * Defines the maximum size in MB of a tile used to decode image when the flag `SDWebImageScaleDownLargeImages` is set
- * Suggested value for iPad1 and iPhone 3GS: 20.
- * Suggested value for iPad2 and iPhone 4: 40.
- * Suggested value for iPhone 3G and iPod 2 and earlier devices: 10.
- */
-static const CGFloat kSourceImageTileSizeMB = 20.0f;
+@end
 
-static const CGFloat kBytesPerMB = 1024.0f * 1024.0f;
-static const CGFloat kPixelsPerMB = kBytesPerMB / kBytesPerPixel;
-static const CGFloat kDestTotalPixels = kDestImageSizeMB * kPixelsPerMB;
-static const CGFloat kTileTotalPixels = kSourceImageTileSizeMB * kPixelsPerMB;
+@implementation SDWebImageDecoder
 
-static const CGFloat kDestSeemOverlap = 2.0f;   // the numbers of pixels to overlap the seems where tiles meet.
++ (instancetype)sharedCoder {
```

Comment:
Sounds great. But in fact all the protocol is optional, and our `SDWebImageGIFCoder` do not support `incrementalDecodedImageWithData` this one. If it subclass from that `SDWebImageDecoder` abstract class(which implements all the methods), it maybe called this function. This is not as excepted.

In fact, `SDImageCache` will use the provided coder instance to decode, `SDWebImageDownloader` need to transfer the coder class to actual `SDWebImageDownloadOperation` because each operation need a seperate coder to avoid data conflicts. But maybe there will be a better solution.